File: api/trades.js
import { readFileSync, readdirSync } from 'fs'
import { join, dirname } from 'path'
import { fileURLToPath } from 'url'
import { parse } from 'csv-parse/sync'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const RESULTS_DIR = join(__dirname, '..', 'results')

function readCSV(filename) {
  try {
    const csv = readFileSync(join(RESULTS_DIR, filename), 'utf8')
    return parse(csv, { columns: true, skip_empty_lines: true })
  } catch {
    return []
  }
}

function loadAllTrades() {
  const trades = [...readCSV('trade_log.csv'), ...readCSV('paper_trades.csv')]
  return trades.map(t => ({
    ...t,
    entry_price: parseFloat(t.entry_price) || 0,
    exit_price: parseFloat(t.exit_price) || 0,
    return_pct: parseFloat(t.return_pct) || 0,
    net_pnl: parseFloat(t.net_pnl) || 0,
    position_size: parseFloat(t.position_size) || 0,
    hold_days: parseInt(t.hold_days) || 0,
    shares: parseInt(t.shares) || 0,
  }))
}

export default function handler(req, res) {
  console.log('RESULTS_DIR:', RESULTS_DIR)
  try { console.log('Files:', readdirSync(RESULTS_DIR)) } catch (e) { console.log('readdirSync error:', e.message) }
  res.setHeader('Access-Control-Allow-Origin', '*')
  const url = new URL(req.url, `http://${req.headers.host}`)
  const path = url.pathname.replace(/\/+$/, '')

  if (path.endsWith('/trades/stats')) {
    return handleStats(res)
  }
  if (path.endsWith('/trades/export')) {
    return handleExport(res)
  }
  return handleList(url, res)
}

function handleList(url, res) {
  const page = parseInt(url.searchParams.get('page')) || 1
  const perPage = parseInt(url.searchParams.get('per_page')) || 50
  const ticker = url.searchParams.get('ticker') || ''
  const start = url.searchParams.get('start') || ''
  const end = url.searchParams.get('end') || ''
  const exitReason = url.searchParams.get('exit_reason') || ''

  let trades = loadAllTrades()

  if (ticker) trades = trades.filter(t => (t.ticker || '').toLowerCase().includes(ticker.toLowerCase()))
  if (start) trades = trades.filter(t => (t.entry_date || '') >= start)
  if (end) trades = trades.filter(t => (t.entry_date || '') <= end)
  if (exitReason) trades = trades.filter(t => t.exit_reason === exitReason)

  const total = trades.length
  const pages = Math.ceil(total / perPage)
  const startIdx = (page - 1) * perPage
  const pageTrades = trades.slice(startIdx, startIdx + perPage)

  res.json({ trades: pageTrades, total, page, pages })
}

function handleStats(res) {
  const trades = loadAllTrades()
  if (!trades.length) return res.json({ total_trades: 0 })

  const returns = trades.map(t => t.return_pct)
  const wins = returns.filter(r => r > 0)
  const losses = returns.filter(r => r <= 0)

  const stats = {
    total_trades: trades.length,
    win_rate: trades.length > 0 ? Math.round(wins.length / trades.length * 1000) / 10 : 0,
    avg_win: wins.length > 0 ? Math.round(wins.reduce((a, b) => a + b, 0) / wins.length * 100) / 100 : 0,
    avg_loss: losses.length > 0 ? Math.round(losses.reduce((a, b) => a + b, 0) / losses.length * 100) / 100 : 0,
    best_trade: returns.length > 0 ? Math.round(Math.max(...returns) * 100) / 100 : 0,
    worst_trade: returns.length > 0 ? Math.round(Math.min(...returns) * 100) / 100 : 0,
    avg_hold_days: trades.length > 0 ? Math.round(trades.reduce((s, t) => s + t.hold_days, 0) / trades.length * 10) / 10 : 0,
  }

  const byExitReason = {}
  trades.forEach(t => {
    const reason = t.exit_reason || 'unknown'
    byExitReason[reason] = (byExitReason[reason] || 0) + 1
  })
  stats.by_exit_reason = byExitReason

  res.json(stats)
}

function handleExport(res) {
  try {
    const content = readFileSync(join(RESULTS_DIR, 'trade_log.csv'), 'utf8')
    res.setHeader('Content-Type', 'text/csv')
    res.setHeader('Content-Disposition', 'attachment; filename=trade_log.csv')
    res.send(content)
  } catch {
    res.json({ error: 'No trade log found' })
  }
}

File: api/trades.py
"""Vercel serverless: /api/trades — trade history with filtering and pagination."""

from http.server import BaseHTTPRequestHandler
import json
import csv
import os
from urllib.parse import urlparse, parse_qs

RESULTS_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'results')


def _read_csv(filename):
    path = os.path.join(RESULTS_DIR, filename)
    if not os.path.exists(path):
        return []
    with open(path, newline='', encoding='utf-8') as f:
        return list(csv.DictReader(f))


def _safe_float(val, default=0.0):
    try:
        return float(val)
    except (ValueError, TypeError):
        return default


def _load_all_trades():
    trades = []
    for name in ['trade_log.csv', 'paper_trades.csv']:
        trades.extend(_read_csv(name))
    # Convert numeric fields
    for t in trades:
        for key in ['entry_price', 'exit_price', 'return_pct', 'net_pnl', 'position_size']:
            if key in t:
                t[key] = _safe_float(t[key])
        for key in ['hold_days', 'shares']:
            if key in t:
                try:
                    t[key] = int(float(t[key]))
                except (ValueError, TypeError):
                    t[key] = 0
    return trades


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        parsed = urlparse(self.path)
        params = parse_qs(parsed.query)

        path = parsed.path.rstrip('/')

        if path.endswith('/trades/stats'):
            self._handle_stats()
        elif path.endswith('/trades/export'):
            self._handle_export()
        else:
            self._handle_list(params)

    def _handle_list(self, params):
        page = int(params.get('page', ['1'])[0])
        per_page = int(params.get('per_page', ['50'])[0])
        ticker = params.get('ticker', [''])[0]
        start = params.get('start', [''])[0]
        end = params.get('end', [''])[0]
        exit_reason = params.get('exit_reason', [''])[0]

        trades = _load_all_trades()

        if ticker:
            trades = [t for t in trades if ticker.lower() in t.get('ticker', '').lower()]
        if start:
            trades = [t for t in trades if t.get('entry_date', '') >= start]
        if end:
            trades = [t for t in trades if t.get('entry_date', '') <= end]
        if exit_reason:
            trades = [t for t in trades if t.get('exit_reason', '') == exit_reason]

        total = len(trades)
        pages = (total + per_page - 1) // per_page if per_page > 0 else 0
        start_idx = (page - 1) * per_page
        page_trades = trades[start_idx:start_idx + per_page]

        data = {"trades": page_trades, "total": total, "page": page, "pages": pages}
        self._json_response(data)

    def _handle_stats(self):
        trades = _load_all_trades()
        if not trades:
            self._json_response({"total_trades": 0})
            return

        returns = [t.get('return_pct', 0) for t in trades]
        wins = [r for r in returns if r > 0]
        losses = [r for r in returns if r <= 0]

        stats = {
            "total_trades": len(trades),
            "win_rate": round(len(wins) / len(trades) * 100, 1) if trades else 0,
            "avg_win": round(sum(wins) / len(wins), 2) if wins else 0,
            "avg_loss": round(sum(losses) / len(losses), 2) if losses else 0,
            "best_trade": round(max(returns), 2) if returns else 0,
            "worst_trade": round(min(returns), 2) if returns else 0,
            "avg_hold_days": round(
                sum(t.get('hold_days', 0) for t in trades) / len(trades), 1
            ) if trades else 0,
        }

        exit_reasons = {}
        for t in trades:
            reason = t.get('exit_reason', 'unknown')
            exit_reasons[reason] = exit_reasons.get(reason, 0) + 1
        stats["by_exit_reason"] = exit_reasons

        self._json_response(stats)

    def _handle_export(self):
        path = os.path.join(RESULTS_DIR, 'trade_log.csv')
        if os.path.exists(path):
            with open(path, 'r', encoding='utf-8') as f:
                content = f.read()
            self.send_response(200)
            self.send_header('Content-Type', 'text/csv')
            self.send_header('Content-Disposition', 'attachment; filename=trade_log.csv')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(content.encode())
        else:
            self._json_response({"error": "No trade log found"})

    def _json_response(self, data):
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())
